lib/indexer/change_notification_processor.rb

```diff
@@ -0,0 +1,54 @@
+# ChangeNotificationProcessor
+#
+# This is called by the message queue processor and is responsible
+# for acting on the incoming payload. It checks if the updated document in the
+# payload needs to be updated, looks up the document in elasticsearch and triggers
+# a "re-index", which means that the document will go through the DocumentPreparer
+# process and will look up the tags for the document again.
+module Indexer
+  class ChangeNotificationProcessor
+    EXCLUDED_FORMATS = %w{email_alert_signup gone redirect}.freeze
+
+    # Gets initialised with a content-item from the publishing-api message queue.
+    def initialize(content_item)
+      @content_item = content_item
+    end
+
+    def trigger
+      return if should_skip_document?
+      trigger_indexing_of_document
+    rescue KeyError => e
+      raise Indexer::MalformedMessage, "Content item attribute missing. #{e.message}"
+    end
+
+  private
+
+    attr_reader :content_item
+
+    def document
+      @document ||= begin
```

Comment:
I think if we extracted this it would basically be the same method as `document` anyway, minus the memoisation. It's not something I care too much about when calling it.